### PR TITLE
Feature/tvist1 743 copy case event

### PR DESCRIPTION
**Pull request by jekuaitk**
https://jira.itkdev.dk/browse/TVIST1-743
https://jira.itkdev.dk/browse/TVIST1-789

* Adds ability to copy case events
* Adds [select2](https://select2.org/) to copy documents and case events.
* Updates `CaseEvent` `CaseEntity` relation to `manyToMany`.

### CODE REVIEW

File: templates/case/event/copy.html.twig [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{% endblock %}
+
+{% block content %}
+    <main role="main" class="col-12 mt-3">
+        {{ form_start(form) }}
+        <div>
+            <label >{% trans %}Cases{% endtrans %}</label>
+            {{ form_widget(form.cases) }}
+        </div>
+        <button class="btn btn-success float-right mt-3">{{ button_label|default('Copy'|trans([], 'case_event', 'da')) }}</button>
```
Comment (rimi-itk) [inline templates/case/event/copy.html.twig:25]: Why are you passing a locale (`'da'`) here?

File: templates/case/event/copy.html.twig [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        {{ form_end(form) }}
+    </main>
+{% endblock %}
+
+{% block javascripts %}
+    {{ parent() }}
+    {{ encore_entry_script_tags('case_event_copy_select2') }}
```
Comment (rimi-itk) [inline templates/case/event/copy.html.twig:32]: `_select2` should not be part of this name.

File: webpack.config.js [hunk 3]
```diff
@@ -38,6 +38,8 @@ Encore
   .addEntry('case_new_prepare', './assets/case/municipality_and_board_selector.js')
   .addEntry('mail_templates_custom_data', './assets/mail_templates/custom_data.js')
   .addEntry('document_case_event', './assets/document/case_event.js')
+  .addEntry('case_event_copy_select2', './assets/case_event/copy_select2.js')
+  .addEntry('document_copy_select2', './assets/document/copy_select2.js')
```
Comment (rimi-itk) [inline webpack.config.js:42]: “_select2” should not be a part of these names. We don't want to rename if we switch to https://tom-select.js.org/, say. I recommend switching to Tom Select (which, by the way, is used by Easy Admin).

File: src/Form/CaseEventCopyType.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        $suitableCases = $options['suitableCases'];
+
+        $builder->add('cases', EntityType::class, [
+            'choices' => $suitableCases,
+            'class' => get_class($case),
+            'label' => $this->translator->trans('Copy to', [], 'case_event'),
+            'multiple' => true,
+            'attr' => [
+                'class' => 'select2',
+                'data-placeholder' => $this->translator->trans('Click to select', [], 'case_event'),
```
Comment (rimi-itk) [inline src/Form/CaseEventCopyType.php:39]: “Click to select” doesn't really say anything about what to do (apart from clicking, but that's not really informative; I could also use the keyboard to select something). “Select cases” will make more sense.

File: src/Form/CopyDocumentForm.php [hunk 5]
```diff
@@ -27,10 +32,11 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         $builder->add('cases', EntityType::class, [
             'choices' => $suitableCases,
             'class' => get_class($case),
-            'label' => 'Copy to',
+            'label' => $this->translator->trans('Copy to', [], 'documents'),
             'multiple' => true,
             'attr' => [
-                'size' => 10,
+                'class' => 'select2',
+                'data-placeholder' => $this->translator->trans('Click to select', [], 'documents'),
```
Comment (rimi-itk) [inline src/Form/CopyDocumentForm.php:39]: “Click to select” is a somewhat unorthodox placeholder text. “Select cases” will make more sense.

File: assets/case_event/copy_select2.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+/* global $ */
+
+window.addEventListener('load', function () {
+  const $selector = $('#case_event_copy_cases')
```
Comment (rimi-itk) [inline assets/case_event/copy_select2.js:4]: Consider using a more generic selector, here, e.g. `select.select2` or `select.select-autocomplete`, say.

File: src/Service/CaseEventHelper.php [hunk 7]
```diff
@@ -85,4 +88,44 @@ private function getLines(?string $additionalParties): array
     {
         return array_filter(array_map('trim', explode(PHP_EOL, $additionalParties)));
     }
+
+    public function handleCopyCaseEventForm(array $cases, CaseEvent $caseEvent)
```
Comment (rimi-itk) [inline src/Service/CaseEventHelper.php:92]: “handleCopyCaseEventForm” does really tell what this function does. `copyCaseEventToCases` may be a better name.

File: templates/case/event/copy.html.twig [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        <div class="col-auto">
+            <a class="btn btn-primary" href="{{ path('case_event_show', {id: case.id, caseEvent: case_event.id}) }}">{% trans %}Cancel{% endtrans %}</a>
+        </div>
+    </div>
+{% endblock %}
+
+{% block content %}
+    <main role="main" class="col-12 mt-3">
+        {{ form_start(form) }}
+        <div>
```
Comment (rimi-itk) [inline templates/case/event/copy.html.twig:21]: Maybe we should tell the user that the case events are not actually copied, but a linked to other cases, i.e. changing a case event will change it on all (linked) cases.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: src/Repository/CaseEventRepository.php [hunk 9]
```diff
@@ -32,7 +32,8 @@ public function findByCase(CaseEntity $case, array $criteria = [], array $orderB
     public function createAvailableCaseEventsForCaseQueryBuilder($alias, CaseEntity $caseEntity): QueryBuilder
     {
         return $this->createQueryBuilder($alias)
-            ->where($alias.'.caseEntity = :case')
+            ->join($alias.'.caseEntities', 'hest')
+            ->where('hest.id = :case')
```
[the remark this answers is not in the document]
Comment (jekuaitk) [reply src/Repository/CaseEventRepository.php:36]: Woops. Updated.

File: templates/case/event/copy.html.twig (hunk 1 above)
Comment (jekuaitk) [reply templates/case/event/copy.html.twig:25]: Removed.

File: src/Service/CaseEventHelper.php (hunk 7 above)
Comment (jekuaitk) [reply src/Service/CaseEventHelper.php:92]: Renamed.

File: templates/case/event/copy.html.twig (hunk 2 above)
Comment (jekuaitk) [reply templates/case/event/copy.html.twig:32]: Removed.

File: src/Service/CaseEventHelper.php (hunk 7 above)
Comment (jekuaitk) [reply src/Service/CaseEventHelper.php:92]: Renamed.

File: src/Form/CaseEventCopyType.php (hunk 4 above)
Comment (jekuaitk) [reply src/Form/CaseEventCopyType.php:39]: 👍 Reworded.

File: src/Form/CopyDocumentForm.php (hunk 5 above)
Comment (jekuaitk) [reply src/Form/CopyDocumentForm.php:39]: Reworded.

File: webpack.config.js (hunk 3 above)
Comment (jekuaitk) [reply webpack.config.js:42]: Removed the `_select2` part of the names. 

Lets wait with the change to `tom-select` (https://symfony.com/doc/current/frontend/ux.html).

File: assets/case_event/copy_select2.js (hunk 6 above)
Comment (jekuaitk) [reply assets/case_event/copy_select2.js:4]: Updated selector.

File: templates/case/event/copy.html.twig (hunk 8 above)
Comment (jekuaitk) [reply templates/case/event/copy.html.twig:21]: Added helper texting explaining that we link rather than copy.

**Review by rimi-itk — APPROVED** [review]

**Pull request merged**